pmd-dist/src/test/java/net/sourceforge/pmd/docs/RuleDocGenerator.java

```diff
@@ -0,0 +1,295 @@
+        description = description.replaceAll("\\n|\\r", " ");
+        description = StringUtils.stripToEmpty(description);
+        return description;
+    }
+
+    /**
+     * Generates for each ruleset a page. The page contains the details for each rule.
+     *
+     * @param rulesets all rulesets
+     * @throws IOException
+     */
+    private void generateRuleSetIndex(Map<Language, List<RuleSet>> rulesets) throws IOException {
+        for (Map.Entry<Language, List<RuleSet>> entry : rulesets.entrySet()) {
+            String languageTersename = entry.getKey().getTerseName();
+            for (RuleSet ruleset : entry.getValue()) {
+                String filename = RULESET_INDEX_FILENAME_PATTERN
+                    .replace("${language.tersename}", languageTersename)
+                    .replace("${ruleset.name}", getRuleSetFilename(ruleset));
+
+                Path path = getAbsoluteOutputPath(filename);
+
+                String permalink = RULESET_INDEX_PERMALINK_PATTERN
+                        .replace("${language.tersename}", languageTersename)
+                        .replace("${ruleset.name}", getRuleSetFilename(ruleset));
+
+                List<String> lines = new LinkedList<>();
+                lines.add("---");
+                lines.add("title: " + ruleset.getName());
+                lines.add("summary: " + getRuleSetDescriptionSingleLine(ruleset));
+                lines.add("permalink: " + permalink);
+                lines.add("folder: pmd/rules/" + languageTersename);
+                lines.add("sidebaractiveurl: /" + LANGUAGE_INDEX_PERMALINK_PATTERN.replace("${language.tersename}", languageTersename));
+                lines.add("editmepath: ../" + getRuleSetSourceFilepath(ruleset));
+                lines.add("---");
+                
+                for (Rule rule : ruleset.getRules()) {
+                    lines.add("## " + rule.getName());
+                    if (rule.getSince() != null) {
+                        lines.add("**Since:** " + rule.getSince());
+                        lines.add("");
+                    }
+                    lines.add("**Priority:** " + rule.getPriority() + " (" + rule.getPriority().getPriority() + ")");
+                    lines.add("");
+
+                    lines.add(StringUtils.stripToEmpty(rule.getDescription()));
+                    lines.add("");
+                    if (!rule.getExamples().isEmpty()) {
+                        lines.add("**Example(s):**");
+                        for (String example : rule.getExamples()) {
+                            lines.add("```");
+                            lines.add(StringUtils.stripToEmpty(example));
+                            lines.add("```");
+                            lines.add("");
+                        }
+                    }
+                    if (!rule.getPropertyDescriptors().isEmpty()) {
+                        lines.add("**This rule has the following properties:**");
```

Comment:
We could filter out some properties here, eg `xpath` and `version` for XPath rules; `violationSuppressRegex` and `violationSuppressRegex` since all rules have these two